File: backend/app/api/v1/alerts.py
"""
Alerts API Router  —  /api/v1/alerts
======================================
Handles alert ingestion, retrieval, triage triggering, and status management.
"""

from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, BackgroundTasks, Depends, Path, Query, status
from fastapi.responses import ORJSONResponse

from app.core.logging import get_logger
from app.core.security import AnalystRequired, TokenPayload, ViewerRequired
from app.models.alert import (
    AlertIngest,
    AlertListResponse,
    AlertResponse,
    AlertSource,
    AlertStatus,
    AlertStatusUpdate,
    Severity,
)
from app.models.common import SuccessResponse
from app.services.alert_service import AlertService

logger = get_logger(__name__)
router = APIRouter(prefix="/alerts", tags=["Alerts"])

# ── Dependency: alert service ─────────────────────────────────────────────────

def get_alert_service() -> AlertService:
    return AlertService()


# ── Endpoints ─────────────────────────────────────────────────────────────────

@router.post(
    "/ingest",
    response_model=SuccessResponse[AlertResponse],
    status_code=status.HTTP_202_ACCEPTED,
    summary="Ingest a new security alert",
    description=(
        "Accepts a raw security event, indexes it in Elasticsearch, and "
        "triggers the async enrichment + triage pipeline in the background."
    ),
)
async def ingest_alert(
    payload: AlertIngest,
    background_tasks: BackgroundTasks,
    token: TokenPayload = AnalystRequired,
    svc: AlertService = Depends(get_alert_service),
) -> SuccessResponse[AlertResponse]:
    logger.info(
        "alert_ingest_requested",
        source=payload.source,
        user_id=token.user_id,
    )
    alert = await svc.create_alert(payload, created_by=token.user_id)
    background_tasks.add_task(svc.run_enrichment_pipeline, alert.id)
    return SuccessResponse(
        data=alert,
        message="Alert ingested. Enrichment pipeline started.",
    )


@router.get(
    "/",
    response_model=SuccessResponse[AlertListResponse],
    summary="List alerts",
    description="Paginated, filterable list of security alerts.",
)
async def list_alerts(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=25, ge=1, le=100),
    status: Optional[AlertStatus] = Query(default=None),
    severity: Optional[Severity] = Query(default=None),
    source: Optional[AlertSource] = Query(default=None),
    search: Optional[str] = Query(default=None, description="Full-text search term"),
    token: TokenPayload = ViewerRequired,
    svc: AlertService = Depends(get_alert_service),
) -> SuccessResponse[AlertListResponse]:
    result = await svc.list_alerts(
        page=page,
        page_size=page_size,
        status_filter=status,
        severity_filter=severity,
        source_filter=source,
        search=search,
    )
    return SuccessResponse(data=result)


@router.get(
    "/{alert_id}",
    response_model=SuccessResponse[AlertResponse],
    summary="Get alert by ID",
)
async def get_alert(
    alert_id: str = Path(..., description="Elasticsearch document ID"),
    token: TokenPayload = ViewerRequired,
    svc: AlertService = Depends(get_alert_service),
) -> SuccessResponse[AlertResponse]:
    alert = await svc.get_alert(alert_id)
    return SuccessResponse(data=alert)


@router.post(
    "/{alert_id}/triage",
    response_model=SuccessResponse[AlertResponse],
    status_code=status.HTTP_202_ACCEPTED,
    summary="Trigger AI triage for an alert",
    description=(
        "Manually triggers the LangChain triage agent for the specified alert. "
        "Useful for re-triaging after analyst feedback or enrichment updates."
    ),
)
async def trigger_triage(
    alert_id: str = Path(..., description="Alert ID to triage"),
    background_tasks: BackgroundTasks = BackgroundTasks(),
    token: TokenPayload = AnalystRequired,
    svc: AlertService = Depends(get_alert_service),
) -> SuccessResponse[AlertResponse]:
    alert = await svc.get_alert(alert_id)
    background_tasks.add_task(svc.run_triage, alert_id, triggered_by=token.user_id)
    logger.info("triage_manually_triggered", alert_id=alert_id, user_id=token.user_id)
    return SuccessResponse(
        data=alert,
        message="Triage pipeline triggered. Results will appear shortly.",
    )


@router.put(
    "/{alert_id}/status",
    response_model=SuccessResponse[AlertResponse],
    summary="Update alert status",
)
async def update_alert_status(
    payload: AlertStatusUpdate,
    alert_id: str = Path(...),
    token: TokenPayload = AnalystRequired,
    svc: AlertService = Depends(get_alert_service),
) -> SuccessResponse[AlertResponse]:
    alert = await svc.update_status(
        alert_id=alert_id,
        new_status=payload.status,
        comment=payload.comment,
        updated_by=token.user_id,
    )
    logger.info(
        "alert_status_updated",
        alert_id=alert_id,
        new_status=payload.status,
        user_id=token.user_id,
    )
    return SuccessResponse(data=alert, message=f"Alert status updated to {payload.status.value}.")


@router.delete(
    "/{alert_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete an alert",
    description="Permanently removes an alert. Restricted to admin role.",
)
async def delete_alert(
    alert_id: str = Path(...),
    token: TokenPayload = AnalystRequired,
    svc: AlertService = Depends(get_alert_service),
) -> None:
    await svc.delete_alert(alert_id)
    logger.info("alert_deleted", alert_id=alert_id, user_id=token.user_id)

File: frontend/src/mocks/handlers/alerts.ts
/**
 * SentinelAI — MSW Alert Handlers
 *
 * Mocks: GET /api/v1/alerts, GET /api/v1/alerts/:id,
 *        PUT /api/v1/alerts/:id/status, POST /api/v1/alerts/:id/triage
 */

import { http, HttpResponse } from "msw";
import { mockAlerts } from "@/mocks/data/alerts";
import type { AlertListResponse, AlertResponse, AlertStatusUpdatePayload } from "@/types";

const BASE = "/api/v1/alerts";

export const alertHandlers = [
  // ── GET /api/v1/alerts ───────────────────────────────────────────────────
  http.get(BASE + "/", ({ request }) => {
    const url = new URL(request.url);
    const page = parseInt(url.searchParams.get("page") ?? "1", 10);
    const page_size = parseInt(url.searchParams.get("page_size") ?? "25", 10);
    const severity = url.searchParams.get("severity");
    const status = url.searchParams.get("status");
    const source = url.searchParams.get("source");
    const search = url.searchParams.get("search")?.toLowerCase();

    let items = [...mockAlerts];

    if (severity) items = items.filter((a) => a.severity === severity);
    if (status) items = items.filter((a) => a.status === status);
    if (source) items = items.filter((a) => a.source === source);
    if (search) {
      items = items.filter(
        (a) =>
          a.title.toLowerCase().includes(search) ||
          a.description.toLowerCase().includes(search) ||
          a.host?.toLowerCase().includes(search) ||
          a.id.toLowerCase().includes(search)
      );
    }

    const total = items.length;
    const start = (page - 1) * page_size;
    const paged = items.slice(start, start + page_size);

    const data: AlertListResponse = { total, page, page_size, items: paged };

    return HttpResponse.json({
      success: true,
      message: null,
      data,
      request_id: crypto.randomUUID(),
    });
  }),

  // ── GET /api/v1/alerts/:id ───────────────────────────────────────────────
  http.get(`${BASE}/:alertId`, ({ params }) => {
    const alert = mockAlerts.find((a) => a.id === params.alertId);
    if (!alert) {
      return HttpResponse.json(
        { success: false, detail: `Alert '${params.alertId}' not found.` },
        { status: 404 }
      );
    }
    return HttpResponse.json({
      success: true,
      message: null,
      data: alert,
      request_id: crypto.randomUUID(),
    });
  }),

  // ── PUT /api/v1/alerts/:id/status ────────────────────────────────────────
  http.put(`${BASE}/:alertId/status`, async ({ params, request }) => {
    const body = (await request.json()) as AlertStatusUpdatePayload;
    const alert = mockAlerts.find((a) => a.id === params.alertId);
    if (!alert) {
      return HttpResponse.json(
        { success: false, detail: `Alert '${params.alertId}' not found.` },
        { status: 404 }
      );
    }
    const updated: AlertResponse = { ...alert, status: body.status };
    return HttpResponse.json({
      success: true,
      message: `Alert status updated to ${body.status}.`,
      data: updated,
      request_id: crypto.randomUUID(),
    });
  }),

  // ── POST /api/v1/alerts/:id/triage ──────────────────────────────────────
  http.post(`${BASE}/:alertId/triage`, ({ params }) => {
    const alert = mockAlerts.find((a) => a.id === params.alertId);
    if (!alert) {
      return HttpResponse.json(
        { success: false, detail: `Alert '${params.alertId}' not found.` },
        { status: 404 }
      );
    }
    return HttpResponse.json(
      {
        success: true,
        message: "Triage pipeline triggered. Results will appear shortly.",
        data: alert,
        request_id: crypto.randomUUID(),
      },
      { status: 202 }
    );
  }),
];
